CI-mysqlx has never run any test in the last 500 runs (build-cache key never populated)

## Summary

`CI-mysqlx` is **100% red over the last 500 workflow runs**, and has been since at least the API's reach (2026-05-03; likely earlier). **Not a single test in the mysqlx test suite — `test_mysqlx_e2e_*`, `test_mysqlx_route_drop_inflight-t`, the unit-tests subset, the soak harness — has executed in CI in that window.** The entire test family is dead code.

The cause is a build-cache key mismatch in `.github/workflows/CI-mysqlx.yml`: each of its three jobs (`unit-tests`, `e2e-tests`, `soak-tests`) starts with a `Restore build cache` step keyed on `build-<SHA>`, with `fail-on-cache-miss: true`. **No workflow in the repository ever writes that key.** The cache restore therefore fails immediately, and the job aborts before any build, infra setup, or test invocation.

## Evidence

Run history (last 100 runs sampled):

```
100 failure
  0 success
```

500-run lookback for any success — zero matches:
```
gh run list --workflow CI-mysqlx.yml --limit 500 --json conclusion \
  --jq '.[] | select(.conclusion=="success")'
(empty)
```

Latest failure (run [25994509183](https://github.com/sysown/proxysql/actions/runs/25994509183), 2026-05-17 15:09):

```
unit-tests  ##[error]Failed to restore cache entry. Exiting as fail-on-cache-miss is set.
            Input key: build-5044d13d5e33e04dc959dcc81d17313d94205106
```

Other sampled failures (different SHAs, different days), all identical failure mode at `Restore build cache`:

| Run | Date | SHA prefix | Step | Cache key |
|---|---|---|---|---|
| 25994509183 | 2026-05-17 | 5044d13d | Restore build cache | `build-5044d13d…` |
| 25917810000ish | 2026-05-15 | 185abefe | Restore build cache | `build-185abefe…` |
| 25840521367 | 2026-05-14 | 0ceb3368 | Restore build cache | `build-0ceb3368…` |
| 25794062053 | 2026-05-13 | 3726f094 | Restore build cache | `build-3726f094…` |
| 25404686131 | 2026-05-05 | cfad7c85 | Restore build cache | `build-cfad7c85…` |

## Root cause

`.github/workflows/CI-mysqlx.yml` (3 occurrences — lines 30-39, 78-88, 195-202):

```yaml
- name: Restore build cache
  uses: actions/cache@v4
  with:
    path: |
      src/proxysql
      test/
    key: build-${{ github.event.workflow_run && github.event.workflow_run.head_sha || github.sha }}
    # Strict cache dependency when CI-trigger populated the cache
    # upstream; lenient on manual workflow_dispatch so a missing cache
    # doesn't block an ad-hoc run.
    fail-on-cache-miss: ${{ github.event_name == 'workflow_run' }}
```

The inline comment says "Strict cache dependency when CI-trigger populated the cache upstream" — but **CI-trigger doesn't populate any cache** (it's a thin caller that fires `workflow_run` for downstream workflows). The actual cache-writer in the project is `ci-builds.yml` on the `GH-Actions` branch, which uses entirely different cache keys (`<SHA>_<dist><type>_bin` / `<SHA>_<dist><type>_src`, e.g. `9cc20a87…_ubuntu22-tap_bin`). Nothing writes `build-<SHA>`.

`grep -rn 'key: build-' .github/workflows/` returns only the three READ sites in CI-mysqlx.yml. Zero write sites.

## Why this matters

The mysqlx test suite is significant — it covers the chassis plugin loader, the X-Protocol implementation, route management, in-flight survival contracts, the SIGTERM behaviour test scaffolding, the GenAI / MCP / RAG / LLM TAP tests (some of which are `ai-g1`-gated but ultimately route through similar infra). Together that's dozens of test binaries currently un-exercised in CI.

The fact that the workflow has been silently red for 14+ days means real regressions in mysqlx / chassis / X-Protocol code can land on `v3.0` without ever being caught, until a customer hits them in production. This is the same class of failure mode as #5781 / #5794 — "test exists but never actually runs" — except at workflow scale rather than per-test.

Locally, after working around the build-cache and provisioning the missing `alice` user manually, the suite still produces several downstream failures (X-Protocol handshake fails, `setup-infras.bash` has a 10s listener-bind timeout that's racy on a loaded runner). Those are real test/setup issues that should surface in CI once the workflow can actually proceed past `Restore build cache`.

## Suggested fix

Two viable shapes:

### Option A — add an upstream build job to CI-mysqlx that populates the cache

Insert a `build` job in `CI-mysqlx.yml` that runs first and writes the `build-<SHA>` cache:

```yaml
jobs:
  build:
    runs-on: ubuntu-22.04
    if: ${{ github.event.workflow_run && github.event.workflow_run.conclusion == 'success' || ! github.event.workflow_run }}
    steps:
      - uses: actions/checkout@v4
      - name: Save build cache
        uses: actions/cache@v4
        id: build-cache
        with:
          path: |
            src/proxysql
            test/
          key: build-${{ ... head_sha or github.sha ... }}
      - if: steps.build-cache.outputs.cache-hit != 'true'
        run: |
          PROXYSQL40=1 PROXYSQL31=1 PROXYSQLFFTO=1 PROXYSQLTSDB=1 \
            make -j$(nproc) debug && make -j$(nproc) build_tap_test_debug

  unit-tests:
    needs: build
    ...
  e2e-tests:
    needs: build
    ...
  soak-tests:
    needs: build
    ...
```

Pros: self-contained in CI-mysqlx, no cross-workflow coupling. Cons: adds ~5-10 min to first run on each SHA (subsequent jobs hit the warm cache).

### Option B — change the key to match an existing CI-builds output

CI-builds writes `<SHA>_ubuntu22-tap_bin` etc. But that build does NOT use `PROXYSQL40=1`, so the cached `src/proxysql` would not have chassis support and could not load the mysqlx plugin. Would need a new matrix entry in CI-builds that does the PROXYSQL40 build and writes the key CI-mysqlx expects.

I'd recommend **Option A** for the smaller blast radius and clearer ownership.

## After the fix

Once `CI-mysqlx` can actually run its tests, expect a cascade of additional failures to surface that the cache miss has been hiding. Likely categories from local reproduction:

1. **Listener bind timing**: `test/tap/groups/mysqlx-soak/setup-infras.bash` polls for `proxysql:6603` with a 10s deadline. On a loaded CI runner that can race. Either raise the timeout or use exponential backoff.
2. **X-Protocol handshake failures**: locally, even with `alice` provisioned in `mysqlx_users` + `mysql_users` and the route active, all 5 client handshakes fail in `test_mysqlx_route_drop_inflight-t`. Root cause not yet identified — likely TLS-mode mismatch, backend auth setup, or a real plugin bug.
3. **Per-test setup gaps**: `test_mysqlx_route_drop_inflight-t` Stage 3 tries to admin-connect from the test-runner container, but proxysql admin defaults to localhost-only. Either bind admin to the docker network or run that part of the test inside the proxysql container.

These should each get their own issue once the workflow proceeds far enough to surface them.

## Priority

High. The mysqlx code path is a substantial chunk of new-feature surface area in 4.0 that today ships with **zero CI coverage**. Restoring the workflow to a working state is the unblocker for everything else in mysqlx test quality.

--- a/.github/workflows/CI-mysqlx.yml
+++ b/.github/workflows/CI-mysqlx.yml
@@ -12,9 +12,54 @@ concurrency:
   cancel-in-progress: true
 
 jobs:
+  # PROXYSQL40 build of proxysql + TAP tests. The cache write at the end of
+  # this job is what the downstream jobs (unit-tests, e2e-tests, soak-tests)
+  # restore. Before this job existed (see issue #5796), the downstream jobs
+  # all aborted immediately at "Restore build cache" with
+  # `fail-on-cache-miss: true` because no upstream workflow ever wrote the
+  # `build-<SHA>` key — CI-trigger doesn't write any cache, and CI-builds
+  # uses a completely different naming scheme (`<SHA>_<dist><type>_*`). The
+  # net effect was that CI-mysqlx had not executed a single test in the
+  # entire workflow run history (500+ runs, 100% red at the same step).
+  #
+  # This job builds with the matching chassis flags so the cached
+  # `src/proxysql` has the same plugin-services struct layout the downstream
+  # `plugins/mysqlx` build will see — keeping the ABI contract intact.
+  build:
+    if: ${{ github.event.workflow_run && github.event.workflow_run.conclusion == 'success' || ! github.event.workflow_run }}
+    runs-on: ubuntu-22.04
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 1
+
+      - name: Cache build artifacts (restore + save)
+        uses: actions/cache@v4
+        id: build-cache
+        with:
+          path: |
+            src/proxysql
+            test/
+          key: build-${{ github.event.workflow_run && github.event.workflow_run.head_sha || github.sha }}
+
+      - name: Build proxysql (PROXYSQL40 chassis) + TAP tests
+        if: steps.build-cache.outputs.cache-hit != 'true'
+        # The same tier flags the downstream `Build mysqlx plugin` step uses
+        # — see the comment on that step. Mismatch here would let the link
+        # succeed but corrupt memory on the first plugin virtual dispatch.
+        run: |
+          export PROXYSQL40=1
+          export PROXYSQL31=1
+          export PROXYSQLFFTO=1
+          export PROXYSQLTSDB=1
+          make -j$(nproc) debug
+          make -j$(nproc) build_tap_test_debug
+
   unit-tests:
     if: ${{ github.event.workflow_run && github.event.workflow_run.conclusion == 'success' || ! github.event.workflow_run }}
     runs-on: ubuntu-22.04
+    needs: build
     steps:
       - name: Checkout (sparse — test/tap)
         uses: actions/checkout@v4
@@ -33,10 +78,12 @@ jobs:
             src/proxysql
             test/
           key: build-${{ github.event.workflow_run && github.event.workflow_run.head_sha || github.sha }}
-          # Strict cache dependency when CI-trigger populated the cache
-          # upstream; lenient on manual workflow_dispatch so a missing cache
-          # doesn't block an ad-hoc run.
-          fail-on-cache-miss: ${{ github.event_name == 'workflow_run' }}
+          # Strict cache dependency: the `build` job above writes this key
+          # before this job starts (via `needs: build`). A miss here means
+          # the build job failed or the cache was evicted between save and
+          # restore — either way we should fail loudly, not silently
+          # proceed without a binary. See issue #5796.
+          fail-on-cache-miss: true
 
       - name: Build mysqlx plugin
         # PROXYSQL40=1 (and the implied tier flags) MUST match what the
@@ -82,10 +129,13 @@ jobs:
             src/proxysql
             test/
           key: build-${{ github.event.workflow_run && github.event.workflow_run.head_sha || github.sha }}
-          # Strict cache dependency when CI-trigger populated the cache
-          # upstream; lenient on manual workflow_dispatch so a missing cache
-          # doesn't block an ad-hoc run.
-          fail-on-cache-miss: ${{ github.event_name == 'workflow_run' }}
+          # Strict cache dependency: the `build` job at the top of this
+          # workflow writes this key before this job starts (via the
+          # `needs:` chain). A miss means the build job failed or the
+          # cache was evicted between save and restore — either way fail
+          # loudly instead of silently proceeding without a binary. See
+          # issue #5796.
+          fail-on-cache-miss: true
 
       - name: Build mysqlx plugin
         # PROXYSQL40=1 (and the implied tier flags) MUST match what the

